src/boot/efi/random-seed.c

```diff
@@ -221,31 +118,84 @@ static void validate_sha256(void) {
 }
 
 EFI_STATUS process_random_seed(EFI_FILE *root_dir, RandomSeedMode mode) {
-        _cleanup_free_ void *seed = NULL, *new_seed = NULL, *rng = NULL, *for_kernel = NULL, *system_token = NULL;
+        _cleanup_erase_ uint8_t random_bytes[DESIRED_SEED_SIZE], hash_key[HASH_VALUE_SIZE];
+        _cleanup_free_ struct linux_efi_random_seed *new_seed_table = NULL;
+        struct linux_efi_random_seed *previous_seed_table = NULL;
+        _cleanup_free_ void *seed = NULL, *system_token = NULL;
         _cleanup_(file_closep) EFI_FILE *handle = NULL;
-        UINTN size, rsize, wsize, system_token_size = 0;
         _cleanup_free_ EFI_FILE_INFO *info = NULL;
+        _cleanup_erase_ struct sha256_ctx hash;
         uint64_t uefi_monotonic_counter = 0;
+        size_t size, rsize, wsize, i;
+        bool seeded_by_efi = false;
+        uint8_t usage_idx;
         EFI_STATUS err;
 
         assert(root_dir);
+        assert_cc(DESIRED_SEED_SIZE == HASH_VALUE_SIZE);
 
         validate_sha256();
 
         if (mode == RANDOM_SEED_OFF)
                 return EFI_NOT_FOUND;
 
-        /* Let's better be safe than sorry, and for now disable this logic in SecureBoot mode, so that we
-         * don't credit a random seed that is not authenticated. */
-        if (secure_boot_enabled())
-                return EFI_NOT_FOUND;
+        /* hash = LABEL || sizeof(input1) || input1 || ... || sizeof(inputN) || inputN || uefi_monotonic */
+        sha256_init_ctx(&hash);
+
+        /* Some basic domain separation in case somebody uses this data elsewhere */
+        sha256_process_bytes(HASH_LABEL, sizeof(HASH_LABEL) - 1, &hash);
+
+        for (i = 0; i < ST->NumberOfTableEntries; ++i) {
+                if (!memcmp(LINUX_EFI_RANDOM_SEED_TABLE_GUID, &ST->ConfigurationTable[i].VendorGuid,
+                    sizeof(EFI_GUID))) {
+                        previous_seed_table = ST->ConfigurationTable[i].VendorTable;
+                        break;
+                }
+        }
+        if (!previous_seed_table) {
```

Comment:
hmm, doesn't matter at all, but i#d invert this check and put the "good" codepath (i.e. where there's a seed already set) first, and the "fallback" codepath (i.e. where there is not) second. even saves you one "!" byte ;-)

Reply:
The reason I did it this way is lots of other code has the pattern:

```
r = something();
if (r < 0) {
    blah;
    return;
}
keep going
other things
`

In this case, there's no `return`, so the "other things" has to be in an else. But it retains the same reading order of error condition being first.

If you feel strongly though I can change it.

Reply:
nah, doesn't matter, i can see your view. I mean, this is how we usually want it too: error handling in the inner code block/indentation level, and main code flow in the outer block/indentation level